### PR TITLE
preprocessor: add missing #pragma once

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
Code cleanup in some header files : 
- add some missing  `#pragma once`
- add a few missing  include guards 

This may be a small improvement to help fixing #14055
It is a rabbit-hole, so this is not a complete fix.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
pragma once seems to be a non-standard extension.
Not sure this is a good idea if there is a chance it does not work on some older platforms.

I assume that something major changed to cause the slowdown so this is probably looking in the wrong place..

**Comment by mattcaswell** [comment]
> pragma once seems to be a non-standard extension.
Not sure this is a good idea if there is a chance it does not work on some older platforms.

@slontis Surely pragmas are supposed to be ignored on platforms where they are not recognised? So I doubt this will break anything?

**Comment by paulidale** [comment]
I have seen compilers that produced _unrecognised pragma_ errors :(  This is contrary to the standard which states:

> Any pragma that is not recognized by the implementation is ignored.

**Comment by FdaSilvaYY** [comment]
@slontis : I didn't try to grep it, but I think that at least half of the header files already have a `#pragma once ` line.
All <module>_err.h are generated with it. 
So far, I didn't see any complaint about this `#pragma`.

File: include/crypto/dsa.h [hunk 1]
```diff
@@ -34,3 +38,8 @@ int dsa_check_pub_key(const DSA *dsa, const BIGNUM *pub_key, int *ret);
 int dsa_check_pub_key_partial(const DSA *dsa, const BIGNUM *pub_key, int *ret);
 int dsa_check_priv_key(const DSA *dsa, const BIGNUM *priv_key, int *ret);
 int dsa_check_pairwise(const DSA *dsa);
+
+# ifdef  __cplusplus
+}
```
Comment (FdaSilvaYY) [inline include/crypto/dsa.h:43]: oups: this `__cplusplus `is orphan

File: include/crypto/asn1.h [hunk 2]
```diff
@@ -7,7 +7,11 @@
  * https://www.openssl.org/source/license.html
  */
 
-#include <openssl/asn1.h>
+#ifndef OSSL_CRYPTO_ASN1_H
```
Comment (paulidale) [inline include/crypto/asn1.h:10]: Are these guards helpful for internal header files?
We shouldn't be including them more than once.

File: include/crypto/dh.h [hunk 3]
```diff
@@ -7,10 +7,18 @@
...
+#ifndef OSSL_CRYPTO_DH_H
+# define OSSL_CRYPTO_DH_H
+# pragma once
+
+# include <openssl/core.h>
+# include <openssl/params.h>
+# include <openssl/dh.h>
+# include "internal/ffc.h"
+
+# ifdef  __cplusplus
```
Comment (paulidale) [inline include/crypto/dh.h:19]: C++ conditionals are definitely not necessary for _internal_ headers.

**Review by paulidale — COMMENTED** [review]
I like the concept but I'm dubious about adding `#ifndef` guards, C++ `externs` and `#pragma once` directives for _internal_ head files.

Thoughts anyone?

**Comment by mspncp** [comment]
> @slontis : I didn't try to grep it, but I think that at least half of the header files already have a `#pragma once ` line.
> All _err.h are generated with it.
> So far, I didn't see any complaint about this `#pragma`.

Those pragmas were added by me during the grand reorganization of the header files (#9333), after doing some compatibility research. In fact, the `pragma once` is not only tolerated but has been widely adopted and is now [supported by many compilers](https://en.wikipedia.org/wiki/Pragma_once#Portability), including gcc and clang.

File: include/crypto/asn1.h (hunk 2 above)
Comment (mspncp) [reply include/crypto/asn1.h:10]: They are not necessary if it's guaranteed that they are included only once, but they don't do any harm either. But we should try to be consistent as much as possible.

A lot of the internal headers already have include guards, and #9333 did a lot to harmonize naming conventions for the internal include guards. Just looking at `crypto ssl providers`, I count 46 include guards.

```
msp@debian:~/src/openssl$ git grep -n 'ifndef.*_H$' crypto ssl providers | wc -l
46
```

Pragmas are used only twice.

```
msp@debian:~/src/openssl$ git grep -n 'pragma once' crypto ssl providers | wc -l
2
```

**Comment by slontis** [comment]
> > @slontis : I didn't try to grep it, but I think that at least half of the header files already have a `#pragma once ` line.
> > All _err.h are generated with it.
> > So far, I didn't see any complaint about this `#pragma`.
> 
> Those pragmas were added by me during the grand reorganization of the header files (#9333), after doing some compatibility research. In fact, the `pragma once` is not only tolerated but has been widely adopted and is now [supported by many compilers](https://en.wikipedia.org/wiki/Pragma_once#Portability), including gcc and clang.

Obviously my grep fu failed in this case..

**Comment by mspncp** [comment]
> I assume that something major changed to cause the slowdown so this is probably looking in the wrong place..

I agree with @slontis. As @t-j-h [already mentioned](https://github.com/openssl/openssl/issues/14055#issuecomment-773246536), a large amount of the delay (in particular the 950% for configuration on windows) is caused by a lot of perl process creations caused by all the new platform-indepent wrappers like `mkdir-p.pl`, `copy.pl`, even `echo.pl`:
```
Configurations/windows-makefile.tmpl:ECHO="$(PERL)" "$(SRCDIR)\util\echo.pl"
```
This, together with the fact that the former libcrypto is built effectively twice, als default and as fips provider, accounts for a lot of the increases. Include header optimizations might have a little effect, but it may not even be noticed.

**Review by paulidale — APPROVED** [review]

File: include/crypto/evp.h [hunk 4]
```diff
@@ -834,9 +838,12 @@ int evp_pkey_ctx_get1_id_prov(EVP_PKEY_CTX *ctx, void *id);
...
-#endif /* !defined(FIPS_MODULE) */
+# endif /* !defined(FIPS_MODULE) */
+
 void evp_method_store_flush(OSSL_LIB_CTX *libctx);
 int evp_set_default_properties_int(OSSL_LIB_CTX *libctx, const char *propq,
                                    int loadconfig);
 
 void evp_md_ctx_clear_digest(EVP_MD_CTX *ctx, int force);
+
+#endif /* !defined(OSSL_CRYPTO_EVP_H) */
```
Comment (mspncp) [inline include/crypto/evp.h:849]:
```suggestion
#endif /* OSSL_CRYPTO_EVP_H */
```

File: include/crypto/cms.h [hunk 5]
```diff
@@ -18,4 +22,6 @@ int cms_signerinfo_get_signing_cert_v2(CMS_SignerInfo *si,
                                        ESS_SIGNING_CERT_V2 **psc);
 int cms_signerinfo_get_signing_cert(CMS_SignerInfo *si,
                                     ESS_SIGNING_CERT **psc);
+# endif /* ndef OPENSSL_NO_CMS */
```
Comment (mspncp) [inline include/crypto/cms.h:25]:
```suggestion
# endif /* OPENSSL_NO_CMS */
```

File: include/crypto/ess.h [hunk 6]
```diff
@@ -89,3 +93,5 @@ struct ESS_signing_cert_v2_st {
     STACK_OF(ESS_CERT_ID_V2) *cert_ids;
     STACK_OF(POLICYINFO) *policy_info;
 };
+
+#endif
```
Comment (mspncp) [inline include/crypto/ess.h:97]:
```suggestion
#endif /* OSSL_CRYPTO_ESS_H */
```

**Review by mspncp — COMMENTED** [review]
I'm not opposed to adding `#endif` comment, but if you do so, please stick to a consistent style.

**Comment by FdaSilvaYY** [comment]
@mspncp : I agree with your suggestions. this point should be added to the style-guide, may be.
I try to use the GH web interface, and once more, the CLA Checker BOT is not agree with that.
I will rebase this branch tonight to fix CLA issue.

**Comment by mspncp** [comment]
The CLA issue seems to be an artifact caused by the GitHub Web Editor (which commits as `GitHub <noreply@github.com>`). It will disappear when I'll autosquash the commits.

```
commit 1ba42bda39d95101e548ab7b21a7bdab1519b93f (HEAD -> gh-14096, github/gh-14096)
Author:     FdaSilvaYY <FdaSilvaYY@users.noreply.github.com>
AuthorDate: Tue Feb 9 13:49:00 2021 +0100
Commit:     GitHub <noreply@github.com>
CommitDate: Tue Feb 9 13:49:00 2021 +0100

    fixup! include/crypto: add a few missing #pragma once directives
    
    Co-authored-by: Matthias St. Pierre <matthias.st.pierre@ncp-e.com>
```

**Comment by mspncp** [comment]
@FdaSilvaYY I didn't see your post before approving. If you want to do the autosquashing yourself, go ahead. But please use `git rebase -i --keep-base` to avoid tree changes.

**Comment by mspncp** [comment]
> I try to use the GH web interface, and once more, the CLA Checker BOT is not agree with that.

A solution would be to check only the Author, not the Committer. Or, alternatively, to accept  `GitHub <noreply@github.com>` as Committer (only). The case of people using the Suggestion feature or the Web Editor might become more common in the future. If it bugs you, would you mind raising an issue for it?

**Comment by FdaSilvaYY** [comment]
> 
> 
> 24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set. Please review the updates and set the label manually.

I just autosquash and fix the CLA issue.

**Comment by FdaSilvaYY** [comment]
> 
> 
> > I try to use the GH web interface, and once more, the CLA Checker BOT is not agree with that.
> 
> A solution would be to check only the Author, not the Committer. Or, alternatively, to accept `GitHub <noreply@github.com>` as Committer (only). The case of people using the Suggestion feature or the Web Editor might become more common in the future. If it bugs you, would you mind raising an issue for it?

Another solution would be to exclude "fixup!" commits  (and other Git action verbs) from CLA checking, may be ?

File: include/openssl/ess.h.in [hunk 7]
```diff
@@ -15,16 +15,18 @@ use OpenSSL::stackhash qw(generate_stack_macros);
...
-extern "C" {
-# endif
 # include <openssl/safestack.h>
 # include <openssl/x509.h>
 # include <openssl/esserr.h>
 
+# ifdef  __cplusplus
+extern "C" {
+# endif
+
```
Comment (mspncp) [inline include/openssl/ess.h.in:29]: Good catch!

**Review by mspncp — APPROVED** [review]
_(Ups, I forgot to approve)_

**Comment by mspncp** [comment]
> > > I try to use the GH web interface, and once more, the CLA Checker BOT is not agree with that.
> > 
> > 
> > A solution would be to check only the Author, not the Committer. Or, alternatively, to accept `GitHub <noreply@github.com>` as Committer (only). The case of people using the Suggestion feature or the Web Editor might become more common in the future. If it bugs you, would you mind raising an issue for it?
> 
> Another solution would be to exclude "fixup!" commits (and other Git action verbs) from CLA checking, may be ?

I still believe that making an exception for `GitHub <noreply@github.com>`, but only if it's the committer, not the author, would be the most reliable solution to your problem. Those `fixup!` prefixes are just arbitrary text in the commit message which Git happens to add for your convenience. Let's not overengineer it.

**Comment by mspncp** [comment]
The last force-push was an autosquash without tree changes, so no need to reset the grace period.

**Comment by mspncp** [comment]
Merged to master, thanks!

3a111aadc3 include/internal: add a few missing #pragma once directives
d59068bd14 include/openssl: add a few missing #pragma once directives
80ce21fe1a include/crypto: add a few missing #pragma once directives

**Pull request closed without merging**